### PR TITLE
Add/improve badge counts in new room list

### PR DESCRIPTION
**Pull request by turt2live**
**Based on https://github.com/matrix-org/matrix-react-sdk/pull/4732**
~~Based on https://github.com/matrix-org/matrix-react-sdk/pull/4711~~ Merged
**See https://github.com/matrix-org/matrix-react-sdk/compare/pull/4732/head...pull/4734/head for the real diff**
For https://github.com/vector-im/riot-web/issues/13635

This introduces a new component, NotificationBadge, which takes care of all the counting we don't need to do in different places. The same component file has some helper classes to monitor different states of notification to update the notification badge.

The notification badge simply expects to receive a notification state that updates itself. The caller (in our case, the room tile or sublist) doesn't need to worry about much else - just making sure the notification state has the information it needs. In the case of the room tile, it instantiates it and forgets about it. The sublist instantiates the state and then whenever it updates it dumps the list of rooms into the state for processing.

The styles are a bit weird for the reasons described in the CSS. This also sneaks in an IDLE state for rooms in the room list (rooms without badges aka fully read).

image

### CODE REVIEW

File: src/components/views/rooms/NotificationBadge.tsx [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+        this.forceUpdate(); // notification state changed - update
+    };
+
+    public render(): React.ReactElement {
+        // Don't show a badge if we don't need to
+        if (this.props.notification.color <= NotificationColor.Bold) return null;
+
+        const hasNotif = this.props.notification.color >= NotificationColor.Red;
+        const hasCount = this.props.notification.color >= NotificationColor.Grey;
+        const isEmptyBadge = this.props.allowNoCount && !localStorage.getItem("mx_rl_rt_badgeCount");
```
Comment (jryans) [inline src/components/views/rooms/NotificationBadge.tsx:87]: What's the purpose of `mx_rl_rt_badgeCount`...?

File: src/components/views/rooms/NotificationBadge.tsx [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+    public get color(): NotificationColor {
+        return this._color;
+    }
+
+    private get roomIsInvite(): boolean {
+        return getEffectiveMembership(this.room.getMyMembership()) === EffectiveMembership.Invite;
+    }
+
+    public dispose(): void {
```
Comment (jryans) [inline src/components/views/rooms/NotificationBadge.tsx:139]: Hmm, I feel like we more typically use `destroy` from this concept, what's the reason for a different word now? (Off topic and non-blocking for this PR, but it's here, so I figured I'd ask...)

File: src/components/views/rooms/NotificationBadge.tsx (hunk 1 above)
Comment (turt2live) [reply src/components/views/rooms/NotificationBadge.tsx:87]: Whoops, should have called it out. It's a temporary thing to play with counted badges versus dots.

File: src/components/views/rooms/NotificationBadge.tsx (hunk 2 above)
Comment (turt2live) [reply src/components/views/rooms/NotificationBadge.tsx:139]: Muscle memory, will fix across all prs

Comment (turt2live) [reply src/components/views/rooms/NotificationBadge.tsx]: looks like some it landed already - will fix those in a different PR

**Pull request merged**